Bugfix/issue 1474 dirichlet rng tiny

bob-carpenter opened the pull request:
#### Summary:

Add Marsaglia and Tsang PRNG for Dirichlet with alpha < 1

#### Intended Effect:

Avoid returning NaN values from Dirichlet RNG due to underflow of all gamma variates.

#### How to Verify:

New unit tests.

#### Side Effects:

None.

#### Documentation:

Already exists.

#### Reviewer Suggestions:

Anyone.

syclik commented:
Jenkins, retest this please.

syclik commented:
Jenkins, retest this please.

syclik commented on src/stan/math/prim/mat/prob/dirichlet_rng.hpp:
```diff
@@ -3,38 +3,83 @@
       using boost::variate_generator;
       using boost::gamma_distribution;
+      using boost::random::uniform_real_distribution;
+      using stan::math::log_sum_exp;
```
remove `using stan::math::log_sum_exp` statement.

syclik commented on src/stan/math/prim/mat/prob/dirichlet_rng.hpp:
```diff
@@ -3,38 +3,83 @@
-      for (int i = 0; i < alpha.rows(); i++)
-        y(i) /= sum;
-      return y;
+      return y / y.sum();
```
any reason we don't replace with `return y.mean()`?

bob-carpenter replied:
Didn't I answer this?  It's because we want to return
a simplex, not a scalar.

- Bob

> On Jun 9, 2015, at 9:34 AM, Daniel Lee <notifications@github.com> wrote:
> 
> In src/stan/math/prim/mat/prob/dirichlet_rng.hpp:
> 
> >        }
> > -
> > -      for (int i = 0; i < alpha.rows(); i++)
> > -        y(i) /= sum;
> > -      return y;
> > +      return y / y.sum();
> 
> any reason we don't replace with return y.mean()?
> 
> —
> Reply to this email directly or view it on GitHub.
>

syclik replied:
Yes, you did.
On Jun 9, 2015 5:57 PM, "Bob Carpenter" <notifications@github.com> wrote:

> In src/stan/math/prim/mat/prob/dirichlet_rng.hpp
> <https://github.com/stan-dev/stan/pull/1483#discussion_r32069317>:
>
> >        }
> > -
> > -      for (int i = 0; i < alpha.rows(); i++)
> > -        y(i) /= sum;
> > -      return y;
> > +      return y / y.sum();
>
> Didn't I answer this? It's because we want to return a simplex, not a
> scalar. - Bob
> … <#14dda544865b0ad0_>
>  On Jun 9, 2015, at 9:34 AM, Daniel Lee <notifications@github.com> wrote:
> In src/stan/math/prim/mat/prob/dirichlet_rng.hpp: > } > - > - for (int i =
> 0; i < alpha.rows(); i++) > - y(i) /= sum; > - return y; > + return y /
> y.sum(); any reason we don't replace with return y.mean()? — Reply to this
> email directly or view it on GitHub.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1483/files#r32069317>.
>

syclik commented on src/stan/math/prim/mat/prob/dirichlet_rng.hpp:
```diff
@@ -3,38 +3,83 @@
+     * Return a draw from a Dirichlet distribution with specified
+     * parameters and pseudo-random number generator.
+     *
+     * For prior counts greater than zero, the usual algorithm that
```
Should `zero` be `one`?

syclik commented on src/stan/math/prim/mat/prob/dirichlet_rng.hpp:
```diff
@@ -3,38 +3,83 @@
       for (int i = 0; i < alpha.rows(); i++) {
         variate_generator<RNG&, gamma_distribution<> >
-          gamma_rng(rng, gamma_distribution<>(alpha(i, 0), 1));
+          gamma_rng(rng, gamma_distribution<>(alpha(i, 0), 1e-7));
```
Is the `gamma_rng` function better behaved at `1e-7` vs `1`? Also, do we need to normalize if we're using a different coefficient other than 1?

bob-carpenter commented:
I'm going to reply all at once so as not to spam people.

1.  removed unnecessary include of `stan::math::log_sum_exp`

2.  the doc is right as written --- the prior count boundary is zero, because the parameter is prior count plus one --- I added a clarifiation note in the doc because this point confuses everyone (not that anyone reads our function doc).  

3.  no, gamma is not better behaved at 1e-7 and I do not believe more normalization is necessary. 

4.  we can't use `y.mean()` in place of `y / y.sum()` because the former is a scalar and the latter is a vector.

syclik commented:
Thanks!

On Tue, Jun 9, 2015 at 12:32 PM, Bob Carpenter <notifications@github.com>
wrote:

> I'm going to reply all at once so as not to spam people.
>
>    1.
>
>    removed unnecessary include of stan::math::log_sum_exp
>    2.
>
>    the doc is right as written --- the prior count boundary is zero,
>    because the parameter is prior count plus one --- I added a clarifiation
>    note in the doc because this point confuses everyone (not that anyone reads
>    our function doc).
>    3.
>
>    no, gamma is not better behaved at 1e-7 and I do not believe more
>    normalization is necessary.
>    4.
>
>    we can't use y.mean() in place of y / y.sum() because the former is a
>    scalar and the latter is a vector.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1483#issuecomment-110423893>.
>

syclik commented:
I'm merging this without running through all the tests -- I'm messing with the Jenkins config to handle the math lib change, so it'll fail.

I saw the changes made and it's ok.
